File: content/snippets/cpp/interview_lcci_04_03_list_of_depth.cpp
#include <iostream>
#include <queue>
#include <vector>

struct Node { int v; Node* l; Node* r; Node(int x): v(x), l(nullptr), r(nullptr) {} };

std::vector<std::vector<int>> listOfDepth(Node* root) {
    if (!root) return {};
    std::vector<std::vector<int>> out;
    std::queue<Node*> q; q.push(root);
    while (!q.empty()) {
        int n = static_cast<int>(q.size());
        std::vector<int> level;
        while (n--) {
            Node* cur = q.front(); q.pop();
            level.push_back(cur->v);
            if (cur->l) q.push(cur->l);
            if (cur->r) q.push(cur->r);
        }
        out.push_back(level);
    }
    return out;
}

int main() {
    Node* root = new Node(1);
    root->l = new Node(2); root->r = new Node(3);
    root->l->l = new Node(4); root->l->r = new Node(5);
    root->r->l = new Node(6); root->r->r = new Node(7);
    auto levels = listOfDepth(root);
    for (size_t i = 0; i < levels.size(); ++i) {
        if (i) std::cout << " | ";
        for (size_t j = 0; j < levels[i].size(); ++j) {
            if (j) std::cout << ' ';
            std::cout << levels[i][j];
        }
    }
    std::cout << '\n';
}

File: content/snippets/typescript/interview_lcci_04_03_list_of_depth.ts
class Node {
  constructor(
    public val: number,
    public left: Node | null = null,
    public right: Node | null = null
  ) {}
}

function listOfDepth(root: Node | null): number[][] {
  if (!root) return [];
  const out: number[][] = [];
  const q: Node[] = [root];
  while (q.length) {
    const n = q.length;
    const level: number[] = [];
    for (let i = 0; i < n; i++) {
      const cur = q.shift()!;
      level.push(cur.val);
      if (cur.left) q.push(cur.left);
      if (cur.right) q.push(cur.right);
    }
    out.push(level);
  }
  return out;
}

const root = new Node(1, new Node(2, new Node(4), new Node(5)), new Node(3, new Node(6), new Node(7)));
console.log(listOfDepth(root).map((lvl) => lvl.join(" ")).join(" | "));
